addons/web/controllers/json2.py

```diff
@@ -0,0 +1,148 @@
+            or request.httprequest.content_length == 0
+        )
+
+    def dispatch(self, endpoint, args):
+        self.request.params = dict(args)
+        if self.request.httprequest.content_length:
+            try:
+                self.request.params |= self.request.get_json_data()
+            except ValueError as exc:
+                e = f"could not parse the body as json: {exc.args[0]}"
+                raise BadRequest(e) from exc
+            except TypeError as exc:
+                e = "could not parse the body, expecting a json object"
+                raise BadRequest(e) from exc
+        if self.request.db:
+            result = self.request.registry['ir.http']._dispatch(endpoint)
+        else:
+            result = endpoint(**self.request.params)
+        if isinstance(result, http.Response):
+            return result
+        return self.request.make_json_response(result)
+
+    def handle_error(self, exc: Exception) -> Callable:
+        headers = None
+        cookies = None
+        if isinstance(exc, UserError):
+            msg = exc.args[0]
+            status = exc.http_code
+        elif isinstance(exc, HTTPException):
+            if res := exc.response:
+                ct = res.headers.get('Content-Type', 'application/octet-stream')
+                if 'json' in ct or not ct.startswith('text/'):
+                    return exc
+                msg = res.get_data(as_text=True)
+                status = exc.code
+                headers = res.headers
+            else:
+                msg = exc.description
+                status = exc.code
+        elif isinstance(exc, http.SessionExpiredException):
+            # needed by auth='user' but not by auth='bearer'
+            msg = SESSION_EXPIRED_MSG
+            status = HTTPStatus.FORBIDDEN
+        else:
+            msg = "internal server error"
+            status = HTTPStatus.INTERNAL_SERVER_ERROR
+        return self.request.make_json_response(msg, headers, cookies, status)
+
+
+class WebJson2Controller(http.Controller):
+    def _web_json_2_rpc_readonly(self, rule, args):
+        try:
+            model_name = args['model']
+            method_name = args['method']
+            Model = request.registry[model_name]
+        except KeyError:
+            # no need of a read/write cursor to send a 404 http error
+            return True
+        for cls in Model.mro():
+            method = getattr(cls, method_name, None)
+            if method is not None and hasattr(method, '_readonly'):
+                return method._readonly
+        return False
+
+    @http.route(
+        '/json/2/<model>/<method>',
+        # methods=['POST'], # must accept all methods, otherwise /json/<path:subpath> takes over
+        auth='bearer',
+        type='/json/2/rpc',
+        readonly=_web_json_2_rpc_readonly,
+        save_session=False,
+    )
+    def web_json_2_rpc(
+        self,
+        model: str,
+        method: str,
+        ids: Sequence[int] = (),
```

Comment:
Not only integers are possible.
```suggestion
        ids: Sequence[IdType] = (),
```

Reply:
> `IdType: typing.TypeAlias = int | NewId | str`

NewId isn't possible in this case, and typing it with `str` is only gonna confuse people
I'm not sure we would benefit from changing the type.